### PR TITLE
Schema support for the SQL public API

### PR DESCRIPTION
**Pull request by devozerov**
This PR introduces support for schemas to the SQL public API. This is needed for users and the JDBC driver.

We already had all the necessary infrastructure and tests for schemas. However, we never exposed it to the public API. This PR fixes that deficiency. The new message `execute2` is added to the protocol, with the new `schema` field. The schema is then propagated to the optimizer where that actual object name resolution happens. 

The client protocol is adjusted accordingly, see https://github.com/hazelcast/hazelcast-client-protocol/pull/360

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/3123/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/3123/Hazelcast-pr-builder-3123.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-sql-core/src/test/java/com/hazelcast/sql/SqlErrorAbstractTest.java:[282,25] error: execute(HazelcastInstance,SqlStatement) in SqlErrorAbstractTest cannot override execute(HazelcastInstance,SqlStatement) in SqlTestSupport
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:testCompile (default-testCompile) on project hazelcast-sql-core: Compilation failure
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-sql-core/src/test/java/com/hazelcast/sql/SqlErrorAbstractTest.java:[282,25] error: execute(HazelcastInstance,SqlStatement) in SqlErrorAbstractTest cannot override execute(HazelcastInstance,SqlStatement) in SqlTestSupport
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-sql-core/src/test/java/com/hazelcast/sql/SqlErrorAbstractTest.java:[282,25] error: execute(HazelcastInstance,SqlStatement) in SqlErrorAbstractTest cannot override execute(HazelcastInstance,SqlStatement) in SqlTestSupport
--------------------------
</details>
</pre>

**Comment by devozerov** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java:169:17: '100L' is a magic number. [MagicNumber]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java:170:17: '100' is a magic number. [MagicNumber]
--------------------------
</pre>
</details>

**Review by viliam-durina — APPROVED** [review]

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   CodecHashTest.testCodecHashCodes:53->testCodecsInternal:78 ClientRemoveMigrationListenerCodec.java modified by hand. Instead, use hazelcast-client-protocol to generate this codec. Cause: Hash code does not match with the file content. expected:<[c84c567f5516a78187bbeec97c6592de]> but was:<[18dcea7b7ff0158387ad6af967fabd88]>
[INFO] 
[ERROR] Tests run: 35416, Failures: 1, Errors: 0, Skipped: 979
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

**Review by sancar — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/sql/SqlStatement.java [hunk 1]
```diff
@@ -95,6 +97,27 @@ public SqlStatement setSql(@Nonnull String sql) {
...
+        return schema;
+    }
+
+    /**
+     * Sets the current schema name.
+     *
+     * @param schema the current schema name
+     * @return this instance for chaining
+     */
+    public SqlStatement setSchema(String schema) {
```
[the remark this answers is not in the document]
Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/SqlStatement.java]: Applied

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   CodecHashTest.testCodecHashCodes:53->testCodecsInternal:78 ClientRemoveMigrationListenerCodec.java modified by hand. Instead, use hazelcast-client-protocol to generate this codec. Cause: Hash code does not match with the file content. expected:<[c84c567f5516a78187bbeec97c6592de]> but was:<[18dcea7b7ff0158387ad6af967fabd88]>
[INFO] 
[ERROR] Tests run: 35416, Failures: 1, Errors: 0, Skipped: 979
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

**Review by sancar — APPROVED** [review]

**Pull request merged**